### PR TITLE
nsc-events-nextjs_2_183-move-api-call-to-separate-utility-file

### LINKED ISSUE #183: Signup Page: Move API call to separate utility file

### LINKED ISSUE #185: Signup Page: Add error messages if signup fails

### LINKED ISSUE #186: Signup Page: Add API call to check for duplicate  email address

### PR DESCRIPTION
**Pull request by Exochos**
This moves the API call to an external .tsx file. 
I am doing this to help with:
maintainability - this will be easier to read and change.
reusability - we can reuse this component in other forms.
testability - we can test this validation logic much easier using a separate file.
simplicity - this makes both pages easier to read and understand at a glance.

OF NOTE:
 The URL is currently hardcoded to my test server. It should be changed to either localhost or our cloud server address. 
Snackbar messages have been added to indicate success or failure and give the user feedback.
I used this: window.location.href = "/"; to redirect to the home page as I could not get the router working, this also should be changed as it causes a full page refresh. 
Lastly we are still using role: enum in the server side which should be changed to a protected route and an open route, the open route shouldn't have the option of choosing a role otherwise this is a pretty biggly security hole.

It is also industry best practice to abstract and move utility functions to their own files.
Resolves #186 Resolves #183 Resolves #185 

Success

Already

### CODE REVIEW

**Review by heosman — APPROVED** [review]
Looks good. I get a "Sign up successful" message when I sign up, and an error message when I attempt to sign up with the same email, just like the screenshots.

File: app/auth/sign-up/signupApi.tsx [hunk 1]
```diff
@@ -0,0 +1,61 @@
+const URL = "http://159.223.203.135:3000/api/auth/signup";
```
Comment (tinpham5614) [inline app/auth/sign-up/signupApi.tsx:1]: Should we change to localhost:3000?
This doesn't connect to the backend and the database.

File: app/auth/sign-up/page.tsx [hunk 2]
```diff
@@ -70,41 +83,27 @@ const SignUp = () => {
       return;
     }
 
-    // TODO: handle role better - probably need a different route for users vs admins
+    // Generate payload for sign up
     const payload = {
       name: firstName + " " + lastName,
```
Comment (tinpham5614) [inline app/auth/sign-up/page.tsx:88]: name should be separated into firstName, lastName
image

**Review by tinpham5614 — CHANGES_REQUESTED** [review]
Tested it but the new account did not go through the database.

File: app/auth/sign-up/page.tsx (hunk 2 above)
Comment (Exochos) [reply app/auth/sign-up/page.tsx:88]: image

Currently the backend uses a single name field. Refactoring the backend is beyond the scope of this PR

File: app/auth/sign-up/signupApi.tsx (hunk 1 above)
Comment (Exochos) [reply app/auth/sign-up/signupApi.tsx:1]: This connects to a live DB but I can change it to localhost if you prefer
image

File: app/auth/sign-up/signupApi.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+interface SignUpPayload {
+  name: string;
```
[the remark this answers is not in the document]
Comment (Exochos) [reply app/auth/sign-up/signupApi.tsx:3]: image

**Comment by Exochos** [comment]
> Tested it but the new account did not go through the database.

This yours?
image

File: app/auth/sign-up/page.tsx (hunk 2 above)
Comment (tinpham5614) [reply app/auth/sign-up/page.tsx:88]: > Currently the backend uses a single name field. Refactoring the backend is beyond the scope of this PR
The backend was refactored 2 weeks ago.

**Comment by tinpham5614** [comment]
> This yours? 

No, it's not mine. By the way, I can get the sign-up successful message even though I didn't run the backend.

**Comment by Exochos** [comment]
> > This yours?
> 
> No, it's not mine. By the way, I can get the sign-up successful message even though I didn't run the backend.

That is because it is connecting to a live version of the backend server running here: 
http://159.223.203.135:3000
See for example: 
http://159.223.203.135:3000/api/events

I can change it to localhost but this way was easier for testing.

File: app/auth/sign-up/page.tsx (hunk 2 above)
Comment (Exochos) [reply app/auth/sign-up/page.tsx]: I have changed the fields to reflect that change.

**Comment by tinpham5614** [comment]
> That is because it is connecting to a live version of the backend server running here: http://159.223.203.135:3000 See for example: http://159.223.203.135:3000/api/events
> 
> I can change it to localhost but this way was easier for testing.
> 
> EDIT: I have changed it to localhost

It makes sense to me now. Good to know that the live server is ongoing.

**Review by tinpham5614 — APPROVED** [review]

**Pull request merged**